### PR TITLE
Support the remix workflow

### PR DESCRIPTION
**Pull request by gideonthomas**
This is pretty much done!

Although, I do not like the hacks that I need to do for LESS files! Any suggestions (that won't take a lot of work to implement :P) are welcome

### CODE REVIEW

File: api/classes/publisher.js [hunk 1]
```diff
@@ -1,26 +1,13 @@
...
-  });
-} else {
-  client = noxmox.nox.createClient({
-    key: process.env.AWS_ACCESS_KEY_ID,
-    secret: process.env.AWS_SECRET_ACCESS_KEY,
-    bucket: process.env.AWS_BUCKET
-  });
-}
+var client = require('../../lib/amazon-client').create();
+var REMIX_ENDPOINT = Path.join(process.env.REMIX_ENDPOINT, '/');
```
Comment (sedge) [inline api/classes/publisher.js]: Why not just have the env variable include the trailing slash?

File: api/modules/publishedProjects/controller.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+var controller = new BaseController(Model);
+
+function copyFile(file, projectId) {
+  return File.forge({
+    path: file.get('path'),
+    project_id: projectId,
+    buffer: file.get('buffer')
+  }).save()
+  .then(function(projectFile) {
+    if(!projectFile) { throw Boom.notFound(); }
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: Is this block necessary?  AFAIK If the save fails an error will be thrown automatically. 

Likewise, if you don't want the projectfile returned to the parent promise, just:

```js
return Promise.resolve().then(function() {
  File.forge({}).save();
});
```

File: api/modules/publishedProjects/controller.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  var result = Project.forge({
+    title: publishedProject.get('title'),
+    user_id: user.get('id'),
+    tags: publishedProject.get('tags'),
+    description: publishedProject.get('description'),
+    date_created: now,
+    date_updated: now
+  }).save()
+  .then(function(p) {
+    if (!p) { throw Boom.notFound(); }
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: I'm 99% sure you don't need this.

File: api/modules/publishedProjects/controller.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    project = p;
+    return publishedProject.publishedFiles().query({}).fetchAll();
+  })
+  .then(function(files) {
+    if(!files || !files.length) { throw Boom.badImplementation(); }
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: I'm pretty sure a `fetchAll()` always returns something, so this should just be `if (!files.length)`

File: api/modules/publishedProjects/controller.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+var Boom = require('boom');
+
+var Errors = require('../../classes/errors');
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: Nit: `var errors`

File: api/classes/prerequisites.js [hunk 6]
```diff
@@ -71,27 +71,48 @@ prerequisites.validateOwnership = function() {
...
+ * resource being manipulated or requested.
+ *
+ * @return {Promise}
+ */
+prerequisites.validateOwnership = function() {
+  return {
+    method: function validateOwnership(req, reply) {
+      var resource = req.pre.records.models[0];
+      var authenticatedUser = req.pre.user;
+      var result = Promise.resolve().then(function() {
```
Comment (sedge) [inline api/classes/prerequisites.js]: Nit: add a space above this line

File: api/classes/publisher.js [hunk 7]
```diff
@@ -66,32 +50,42 @@ function getPublishedFiles(publishedProject) {
...
-        resolve();
-      } else {
-        reject('S3 upload returned ' + res.statusCode);
-      }
-    });
+    if(mimeType === 'text/html') {
+      buffer = new Buffer(Remix.inject(buffer.toString(), options));
+    }
+
+    upload(path, buffer, mimeType, resolve, reject);
```
Comment (sedge) [inline api/classes/publisher.js]: Nit: This seems like an unnecessary abstraction to me.

Comment (sedge) [inline api/classes/publisher.js]: As a followup, if you really want to keep it then don't instantiate the promise here. Do it in `upload()` and just return:

```js
return upload(path, buffer, mimeType);
```

File: api/classes/publisher.js [hunk 8]
```diff
@@ -66,32 +50,42 @@ function getPublishedFiles(publishedProject) {
...
+    if (res.statusCode === 200) {
+      log.info('Successfully uploaded ' + path);
+      resolve();
+    } else {
+      reject('S3 upload returned ' + res.statusCode);
+    }
+  });
+}
+
+function uploadFile(file, root, options) {
```
Comment (sedge) [inline api/classes/publisher.js]: Change `options` to be `remixMetadata`

File: api/classes/publisher.js [hunk 9]
```diff
@@ -115,15 +109,15 @@ function deleteFileRemotely(file, root) {
   });
 }
 
-function modifyPublishedFiles(action, project, user, files) {
+function modifyPublishedFiles(action, project, user, files, options) {
```
Comment (sedge) [inline api/classes/publisher.js]: `options` -> `remixMetadata`

File: api/classes/publisher.js [hunk 10]
```diff
@@ -136,6 +130,7 @@ exports.publish = function publish(project) {
 
   var publishedFiles;
   var publishedProject;
+  var remixOptions;
```
Comment (sedge) [inline api/classes/publisher.js]: `remixMetadata`

**Comment by gideonthomas** [comment]
Sorry, this is missing another metadata var that needs to be injected viz. the remix endpoint

File: api/modules/publishedProjects/controller.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+var BaseController = require('../../classes/base_controller');
+var Model = require('./model');
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: nit: add a space above this line

File: api/modules/publishedProjects/controller.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+var Project = require('../projects/model');
+var File = require('../files/model');
+var controller = new BaseController(Model);
```
Comment (sedge) [inline api/modules/publishedProjects/controller.js]: nit: add a space above this line

File: env.dist [hunk 13]
```diff
@@ -24,6 +24,9 @@ export S3_EMULATION=true
 # Endpoint for published projects
 export PUBLIC_PROJECT_ENDPOINT="localhost:8001"
 
+# Endpoint for the remix script
```
Comment (sedge) [inline env.dist]: Can you add some extra comments here explaining that the script is responsible for pulling in any resources needed for the remix interface?

**Comment by sedge** [comment]
r+ w/ a couple of nits. 

Is there a thimble.wm.o patch to go along with this so there's something to inject?

**Comment by gideonthomas** [comment]
Still working on that lol

File: api/modules/files/routes/read.js [hunk 14]
```diff
@@ -12,6 +12,7 @@ module.exports = [{
   config: {
     pre: [
       prereqs.confirmRecordExists(Model, 'param', 'id'),
+      prereq.validateUser(),
```
Comment (sedge) [inline api/modules/files/routes/read.js]: `prereqs`

File: api/modules/files/routes/read.js [hunk 15]
```diff
@@ -29,6 +30,7 @@ module.exports = [{
   config: {
     pre: [
       prereqs.confirmRecordExists(Model, 'param', 'project_id'),
+      prereq.validateUser(),
```
Comment (sedge) [inline api/modules/files/routes/read.js]: and here

File: api/modules/files/routes/update.js [hunk 16]
```diff
@@ -20,6 +20,7 @@ module.exports = [{
     pre: [
       prereqs.trackTemporaryFile(),
       prereqs.confirmRecordExists(Model, 'param', 'id'),
+      prereq.validateUser(),
```
Comment (sedge) [inline api/modules/files/routes/update.js]: and here

**Pull request closed without merging**